capa/features/extractors/dnfile/helpers.py

```diff
@@ -299,22 +316,93 @@ def get_dotnet_unmanaged_imports(pe: dnfile.dnPE) -> Iterator[DnUnmanagedMethod]
+        assert isinstance(nestedclass, dnfile.mdtable.NestedClassRow)
+        nested_class_table[nestedclass.NestedClass.row_index] = nestedclass.EnclosingClass.row_index
+
+    return nested_class_table
 
+def get_typedef_class_table(pe):
+    typedef_class_table = []
+    
+    for rid, typedef in iter_dotnet_table(pe, dnfile.mdtable.TypeDef.number):
+        assert isinstance(typedef, dnfile.mdtable.TypeDefRow)
+        typedef_class_table.append((typedef.TypeName, typedef.TypeNamespace))
+        
+    return typedef_class_table
+    
+def typedef_helper(index, nested_class_table, typedef_class_table, n, name):
+    # Append the current typeref name
+    n.append(name)
+    
+    if nested_class_table[index] in nested_class_table:
+        typedef_helper(nested_class_table[index], nested_class_table, typedef_class_table, n, typedef_class_table[nested_class_table[index]-1][0])
+    else:
+        # Document the root enclosing details
+        n.append(typedef_class_table[nested_class_table[index]-1][0])
+        namespace = typedef_class_table[nested_class_table[index]-1][1]
+        n.append(namespace)
+        
+    return tuple(n[::-1])
+
+def get_typeref_table(pe):
+    typeref_table = []
+    
+    # Used to track values in typeref table
+    for rid, typeref in iter_dotnet_table(pe, dnfile.mdtable.TypeRef.number):
+        assert isinstance(typeref, dnfile.mdtable.TypeRefRow)
+        typeref_table.append((typeref.TypeName, typeref.TypeNamespace, typeref.ResolutionScope.row_index, type(typeref.ResolutionScope.table)))
+        
+    return typeref_table
+
+def typeref_helper(index, typeref_table, n, name):
+    # Append the current typeref name
+    n.append(name)
+    
+    if typeref_table[index - 1][3] == dnfile.mdtable.TypeRef:
+        # Recursively call helper function with enclosing typeref details
+        typeref_helper(typeref_table[index - 1][2], typeref_table, n, typeref_table[index - 1][0])
+    else:
+        # Document the root enclosing details
+        n.append(typeref_table[index - 1][0])
+        namespace = typeref_table[index - 1][1]
+        n.append(namespace)
+        
+    return tuple(n[::-1])
+    
 def get_dotnet_types(pe: dnfile.dnPE) -> Iterator[DnType]:
     """get .NET types from TypeDef and TypeRef tables"""
+    nested_class_table = get_nested_class_table(pe)
+    typedef_class_table = get_typedef_class_table(pe)
+    
     for rid, typedef in iter_dotnet_table(pe, dnfile.mdtable.TypeDef.number):
         assert isinstance(typedef, dnfile.mdtable.TypeDefRow)
 
+        typedef_name = []
+        if rid in nested_class_table:
+            typedef.TypeName = typedef_helper(rid, nested_class_table, typedef_class_table, typedef_name, typedef.TypeName)
+
         typedef_token: int = calculate_dotnet_token_value(dnfile.mdtable.TypeDef.number, rid)
         yield DnType(typedef_token, typedef.TypeName, namespace=typedef.TypeNamespace)
 
+
+    typeref_table = get_typeref_table(pe)
```

Comment:
See my earlier comment about indexing the `TypeDef` table directly. Can we do the same here with the `TypeRef` table to avoid iterating it more than is needed?

Reply:
Changed up the accessing to directly index the `TypeRef` table.